File: spring-security-impl/src/main/java/com/alstn113/security/security/authorization/AuthorizationFilter.java
package com.alstn113.security.security.authorization;

import com.alstn113.security.security.context.Authentication;
import com.alstn113.security.security.context.SecurityContextHolder;
import com.alstn113.security.security.exception.AuthorizationDeniedException;
import jakarta.servlet.FilterChain;
import jakarta.servlet.ServletException;
import jakarta.servlet.ServletRequest;
import jakarta.servlet.ServletResponse;
import jakarta.servlet.http.HttpServletRequest;
import jakarta.servlet.http.HttpServletResponse;
import java.io.IOException;
import lombok.RequiredArgsConstructor;
import org.springframework.web.filter.GenericFilterBean;

@RequiredArgsConstructor
public class AuthorizationFilter extends GenericFilterBean {

    private final AuthorizationManager authorizationManager;

    @Override
    public void doFilter(
            ServletRequest servletRequest,
            ServletResponse servletResponse,
            FilterChain filterChain
    ) throws IOException, ServletException {
        HttpServletRequest request = (HttpServletRequest) servletRequest;
        HttpServletResponse response = (HttpServletResponse) servletResponse;

        Authentication authentication = SecurityContextHolder.getContext().getAuthentication();
        AuthorizationResult result = authorizationManager.authorize(() -> authentication, request);

        if (result != null && !result.isGranted()) {
            throw new AuthorizationDeniedException("접근을 위한 권한이 없습니다.");
        }

        filterChain.doFilter(request, response);
    }
}

File: refresh-token-rotation/refresh-token-rotation/src/main/kotlin/io/github/alstn113/rtr/adapter/in/web/security/filter/AuthorizationFilter.kt
package io.github.alstn113.rtr.adapter.`in`.web.security.filter

import com.fasterxml.jackson.databind.ObjectMapper
import io.github.alstn113.rtr.adapter.`in`.web.common.ApiResponse
import io.github.alstn113.rtr.adapter.`in`.web.security.AccountPrincipal
import io.github.alstn113.rtr.adapter.`in`.web.security.AttributeNames
import jakarta.servlet.FilterChain
import jakarta.servlet.http.HttpServletRequest
import jakarta.servlet.http.HttpServletResponse
import org.springframework.http.HttpMethod
import org.springframework.util.AntPathMatcher
import org.springframework.web.filter.OncePerRequestFilter

class AuthorizationFilter(
    private val objectMapper: ObjectMapper,
    private val allowListPatterns: List<String>,
) : OncePerRequestFilter() {

    private val pathMatcher = AntPathMatcher()

    override fun shouldNotFilter(request: HttpServletRequest): Boolean {
        if (request.method == HttpMethod.OPTIONS.name()) {
            return true
        }
        return allowListPatterns.any { pattern -> pathMatcher.match(pattern, request.requestURI) }
    }

    override fun doFilterInternal(
        request: HttpServletRequest,
        response: HttpServletResponse,
        filterChain: FilterChain,
    ) {
        val principal = request.getAttribute(AttributeNames.AUTHENTICATION_PRINCIPAL) as AccountPrincipal?
        if (principal == null || principal.isAnonymous()) {
            writeErrorResponse(response)
            return
        }

        // 권한 검사(Forbidden 403)가 필요한 경우 추가 가능

        filterChain.doFilter(request, response)
    }

    private fun writeErrorResponse(response: HttpServletResponse) {
        response.contentType = "application/json"
        response.characterEncoding = "UTF-8"
        response.status = HttpServletResponse.SC_UNAUTHORIZED

        val errorResponse = ApiResponse.error(errorMessage = "인증이 필요합니다.")
        val body = objectMapper.writeValueAsString(errorResponse)
        response.writer.write(body)
    }
}